plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/rule/WorkloadGroupRuleRoutingService.java

```diff
@@ -0,0 +1,152 @@
+            if (clusterService.state().metadata().hasIndex(indexName)) {
+                routeRequest(request, executePersistLocally, listener, indexName);
+                return;
+            }
+            createIndex(indexName, new ActionListener<>() {
+                @Override
+                public void onResponse(CreateIndexResponse response) {
+                    if (!response.isAcknowledged()) {
+                        logger.error("Failed to create index " + indexName);
+                        listener.onFailure(new IllegalStateException(indexName + " index creation not acknowledged"));
+                    } else {
+                        routeRequest(request, executePersistLocally, listener, indexName);
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    Throwable cause = ExceptionsHelper.unwrapCause(e);
+                    if (cause instanceof ResourceAlreadyExistsException) {
+                        routeRequest(request, executePersistLocally, listener, indexName);
+                    } else {
+                        logger.error("Failed to create index {}: {}", indexName, e.getMessage());
+                        listener.onFailure(e);
+                    }
+                }
+            });
+        }
+    }
+
+    /**
+     * Creates the backing index if it does not exist, then runs the given success callback.
+     * @param indexName the name of the index to create
+     * @param listener listener to handle failures
+     */
+    private void createIndex(String indexName, ActionListener<CreateIndexResponse> listener) {
+        final CreateIndexRequest createRequest = new CreateIndexRequest(indexName).settings(indexSettings);
+        client.admin().indices().create(createRequest, listener);
+    }
+
+    /**
+     * Routes the CreateRuleRequest to the primary shard node for the given index.
+     * Executes locally if the current node is the primary.
+     * @param request the CreateRuleRequest
+     * @param listener listener to handle response or failure
+     * @param indexName the index name used to find the primary shard node
+     */
+    private void routeRequest(
+        CreateRuleRequest request,
+        ActionListener<Void> executePersistLocally,
+        ActionListener<CreateRuleResponse> listener,
+        String indexName
+    ) {
+        Optional<DiscoveryNode> optionalPrimaryNode = getPrimaryShardNode(indexName);
+        if (optionalPrimaryNode.isEmpty()) {
+            listener.onFailure(new IllegalStateException("Primary node for index [" + indexName + "] not found"));
+            return;
+        }
+        DiscoveryNode primaryNode = optionalPrimaryNode.get();
+        if (transportService.getLocalNode().equals(primaryNode)) {
+            executePersistLocally.onResponse(null);
+        } else {
```

Comment:
I think we can avoid checking and writing our redirectlion logic and use else logic every time. `TransportService.sendRequest` internally already checks whether its a local node and redirect accordingly. 

This way we don't require `ActionListener<Void> executePersistLocally` and make it much cleaner, without any overhead.